amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/PeriodicShardSyncManager.java

```diff
@@ -116,24 +153,224 @@ public void stop() {
+            final Map<String, Shard> kinesisShards = shardDetector.listShards().stream()
+                    .collect(Collectors.toMap(Shard::shardId, shard -> shard));
+            return activeLeases.stream().map(lease -> {
+                if (lease.hashKeyRangeForLease() == null) {
+                    final String shardId = lease instanceof MultiStreamLease ?
+                            ((MultiStreamLease) lease).shardId() :
+                            lease.leaseKey();
+                    final Shard shard = kinesisShards.get(shardId);
+                    if(shard == null) {
+                        return lease;
+                    }
+                    lease.hashKeyRange(fromHashKeyRange(shard.hashKeyRange()));
+                    try {
+                        leaseRefresher.updateLease(lease, UpdateField.HASH_KEY_RANGE);
+                    } catch (Exception e) {
+                        log.warn(
+                                "Unable to update hash range key information for lease {} of stream {}. This may result in explicit lease sync.",
+                                lease.leaseKey(), streamIdentifier);
+                    }
+                }
+                return lease;
+            }).filter(lease -> lease.hashKeyRangeForLease() != null).collect(Collectors.toList());
+        } else {
+            return activeLeases;
+        }
+    }
+
+    @VisibleForTesting
+    static Optional<HashRangeHole> checkForHoleInHashKeyRanges(StreamIdentifier streamIdentifier,
+            List<HashKeyRangeForLease> hashKeyRanges, BigInteger minHashKey, BigInteger maxHashKey) {
+        // Sort and merge the overlapping hash ranges.
+        List<HashKeyRangeForLease> mergedHashKeyRanges = sortAndMergeOverlappingHashRanges(hashKeyRanges);
+
+        if(mergedHashKeyRanges.isEmpty()) {
+            log.error("No valid hashranges found for stream {} between {} and {}.", streamIdentifier,
+                    MIN_HASH_KEY, MAX_HASH_KEY);
+            return Optional.of(new HashRangeHole(new HashKeyRangeForLease(MIN_HASH_KEY, MAX_HASH_KEY),
+                    new HashKeyRangeForLease(MIN_HASH_KEY, MAX_HASH_KEY)));
+        }
+
+        // Validate for hashranges bounds.
+        if (!mergedHashKeyRanges.get(0).startingHashKey().equals(minHashKey) || !mergedHashKeyRanges
+                .get(mergedHashKeyRanges.size() - 1).endingHashKey().equals(maxHashKey)) {
+            log.error("Incomplete hash range found for stream {} between {} and {}.", streamIdentifier,
+                    mergedHashKeyRanges.get(0),
+                    mergedHashKeyRanges.get(mergedHashKeyRanges.size() - 1));
+            return Optional.of(new HashRangeHole(mergedHashKeyRanges.get(0),
+                    mergedHashKeyRanges.get(mergedHashKeyRanges.size() - 1)));
+        }
+        // Check for any holes in the sorted hashrange intervals.
+        if (mergedHashKeyRanges.size() > 1) {
```

Comment:
This seems to share a lot of code in common with `EmptyLeaseTableLeaseSynchronizer#hashRangeOfShardsIsComplete`, maybe we can reuse this and move it to another class?

Reply:
it has changed now.